interpreter/cling/lib/Interpreter/LookupHelper.cpp

```diff
@@ -84,14 +86,22 @@ namespace cling {
     //
     //  Create a fake file to parse the type name.
     //
-    std::unique_ptr<llvm::MemoryBuffer>
-      SB = llvm::MemoryBuffer::getMemBufferCopy(code.str() + "\n",
-                                                bufferName.str());
+    FileID FID;
+    llvm::hash_code hashedCode = llvm::hash_value(code);
```

Comment:
How likely are hash value collision?  If I read correctly, the result of a hash collision is 'catastrophic': the LookupHelp search for something and find something *entirely* different ....

Am I missing something?

Reply:
I also think we can store string here too. What do you think @vgvassilev ?

Reply:
The llvm::hash_code is described in [N3333](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3333.html). Based on our observations, the lookuphelper-related memory buffers are less than 100k in rootcling which is the heaviest user of this functionality (partly because of its brokenness due to failing to cache the looked up results). 

If we use sha1 the likelihood of getting an collision is practically 0. See more [here](https://crypto.stackexchange.com/questions/2583/is-it-fair-to-assume-that-sha1-collisions-wont-occur-on-a-set-of-100k-strings). I’d expect similar likelihood for the llvm implementation.

Reply:
Isn't the size actually much smaller than 100k, does the collision risk increase or decrease with smaller buffers?

As importantly do we understand the consequence of a collision (i.e. what it will look like to the user) [worse being silent data corruption, nicest being a nice clear error message that something went wrong]?

Reply:
The collision risk becomes smaller if the count of the memory buffers are less. Moreover, the content of the memory buffers is long enough for a sufficiently smart hash function to avoid colliding. 

If a hash collision happens we will return a wrong FileID and the LookupHelper will parse a memory buffer with different content.

We can probably catch it by comparing the string content of the variable against the content of the SM memory buffer in Debug mode. Yuka, we can simply add an assert somewhere after getting the cached FileID.